### PR TITLE
Enable additional FCRAM

### PR DESCRIPTION
**Pull request by nickdesaulniers**
The LGR models (new 3DS, new 3DS XL, new 2DS XL) all support double the
capacity of FCRAM; 250MiB.

The process in plain English:
1. Interrupt ID #88 is enabled.
2. The clock speed is raised to 804Mhz.
3. The additional FCRAM and L2 cache are enabled.
4. Delay of 403 cycles.
5. Spin in a Wait For Interrupt (WFI) loop until the "SOC mode" changes.
6. Acknowledge the interrupt.
7. Clear the interrupt.

Thanks to @profi200 and the open_agb_firm project which was referenced
extensively.

A change to the DTB file is necessary to inform the kernel that
additional physical memory is available; separate from this commit.

Once enabled, in userspace you can check the amount of physical memory
available via:

$ mount -t proc /proc
$ less /proc/meminfo

From here, we can enable the additional 2 LGR cores, but it's not
required to make use of the additional FCRAM.

See photos at https://github.com/linux-3ds/linux/issues/1#issuecomment-815382458.

### CODE REVIEW

File: arm11/source/smp.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include <stdbool.h>
```
Comment (nickdesaulniers) [inline arm11/source/smp.c:1]: This looks unused now?

**Comment by nickdesaulniers** [comment]
Anecdotally, it felt like this booted much faster. Probably from ~3x clock rate. Oh, I should upload the DTS change, too.

File: common/gic.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+static inline void gic_set_prio(int id, int prio)
+{
+	*(volatile int*)(GIC_DIST_PRI + id) = prio;
+}
+
+static inline void gic_set_target(int id, int core)
+{
+	*(volatile int*)(GIC_DIST_TARGET + id) = core + 1;
+}
```
Comment (nickdesaulniers) [inline common/gic.h:60]: @profi200 I'm not sure I interpreted [this line](https://github.com/profi200/open_agb_firm/blob/1fcf024520571c9e8732a5aaaf1bbd7661cba423/source/arm11/hardware/pdn.c#L70) correctly. How many interrupts can we have? The members of `GicDist` make it look like there's 8*32 possible, but then in your implementation only the first 4*32 get cleared?

File: arm11/source/smp.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	// Write it back to ACK.
+	set_pdn_lgr_socmode(get_pdn_lgr_socmode());
+}
+
+static void set_socmode(void)
+{
+	enable_irqs();
+	enable_fcram_l2();
+	wait_cycles(403);
```
Comment (nickdesaulniers) [inline arm11/source/smp.c:49]: @profi200 what's the purpose of the wait? Is it really necessary?

Comment (profi200) [reply arm11/source/smp.c:49]: Yes. The code in kernel 11 (Horizon OS) does this too. I remember trying to omit this delay and it didn't go well...

The code in open_agb_firm is actually a 1:1 reimplementation of what the original does + a few code quality improvements.

File: common/gic.h (hunk 2 above)
Comment (profi200) [reply common/gic.h:60]: This older GIC (not officially called GIC) supports 256 interrupt lines max. The 3DS uses a maximum of 128.

https://www.3dbrew.org/wiki/ARM11_Interrupts

File: common/arm.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+#ifndef __ASSEMBLY__
+
+#if __ARM_ARM < 6
+static inline void wfi(void)
+{
+	asm volatile (
+		"mov r0, #0\n\t"
+		"mcr p15, 0, r0, c7, c0, 4\n\t");
+}
```
Comment (Wolfvak) [inline common/arm.h:12]: This seems unused right now but just in case it should list r0, cc and memory as clobbered

File: arm11/source/start.s [hunk 5]
```diff
@@ -85,6 +85,8 @@ wait_arm9:
 	ldr r1, =0xFF
 	str r1, [r0, #SCU_INV_ALL_REG]
 
+	bl enable_cores23
```
Comment (Wolfvak) [inline arm11/source/start.s:88]: We really need to setup a sane stack pointer when we start using C (as well as manually clear .bss)

File: common/arm.h (hunk 4 above)
Comment (profi200) [reply common/arm.h:12]: "cc" should not be needed here. It's for when the code modifies flags in the CPSR directly or indirectly. "memory" should be there to prevent reordering by the compiler.

File: common/gic.h (hunk 2 above)
Comment (nickdesaulniers) [reply common/gic.h:60]: > not officially called GIC

heh, perhaps better than the acronym for Distributed Interrupt Controller?  (Thanks again for very helpful links, how do you get there from the main page?)

File: arm11/source/start.s (hunk 5 above)
Comment (nickdesaulniers) [reply arm11/source/start.s:88]: https://github.com/profi200/open_agb_firm/blob/1fcf024520571c9e8732a5aaaf1bbd7661cba423/source/arm11/start.s#L111-L128 looks highly relevant.

Comment (Wolfvak) [reply arm11/source/start.s:88]: Yeah, we don't really need the heap (at least I hope we won't) but stacks are essential because the C runtime straight up assumes it's valid, we've had trouble in the past over that.

If we're only going to be in svc mode, simply allocating some static space in bss and then pointing so to the top of that space should be enough. If we want to handle interrupts (not in the polling way we do rn, but proper interrupts) a lot more setup is needed for the interrupt vectors and memory management in general.

File: common/gic.h (hunk 2 above)
Comment (nickdesaulniers) [reply common/gic.h:60]: ah, I see the wiki has "what links here" links on the left, so I can trace back to the homepage.  I'll add some of these links as comments in the code.

Comment (profi200) [reply common/gic.h:60]: It's a little hidden but i found it by searching for interrupts.

File: arm11/source/start.s (hunk 5 above)
Comment (profi200) [reply arm11/source/start.s:88]: Note that the above code is pretty messy. There are certainly better ways to handle stacks and multiple cores.

Comment (xerpi) [reply arm11/source/start.s:88]: Pretty much the same as the code above:
```c
#define STACKS_BOTTOM   0x.... /* or coming from linker script */
#define STACK_SIZE_LOG2 12
```
```asm
mrc  p15, 0, r0, c0, c0, 5   @ Get CPU ID
ands r0, r0, #3
...
ldr sp, STACKS_BOTTOM
add sp, [r0, lsl #STACK_SIZE_LOG2] @ Stack bottom per core
add sp, #STACK_SIZE_LOG2           @ Stack top per core
```

Comment (Wolfvak) [reply arm11/source/start.s:88]: thankfully secondary cores dont need a stack, their code can all be written in asm for simplicity since it's just a spin table

Comment (nickdesaulniers) [reply arm11/source/start.s:96]: Done. PTAL.

File: common/gic.h (hunk 2 above)
Comment (nickdesaulniers) [reply common/gic.h:60]: I guess I'm still curious though about the comment that says "target core 0" but then assigns a value of 1.  Is the "target" the CPUID plus one?

File: common/arm.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+#if __ARM_ARM < 6
+static inline void wfi(void)
+{
+	asm volatile (
+		"mov r0, #0\n\t"
+		"mcr p15, 0, r0, c7, c0, 4\n\t"::: "r0", "cc", "memory");
+}
+#else
+static inline void wfi(void) { asm volatile ("wfi":::"memory"); }
```
Comment (nickdesaulniers) [inline common/arm.h:14]: as noticed on the stream, this preprocessor guard might not be working as intended.

Comment (nickdesaulniers) [reply common/arm.h:14]: Ah, should be `__ARM_ARCH`.

File: common/gic.h (hunk 2 above)
Comment (profi200) [reply common/gic.h:60]: It's a bitmap. One bit for each CPU core. See here: https://developer.arm.com/documentation/ddi0360/f/mpcore-distributed-interrupt-controller/interrupt-distributor-registers/interrupt-cpu-targets-registers--0x800-0x8fc?lang=en

**Review by Wolfvak — APPROVED** [review]

File: common/gic.h (hunk 2 above)
Comment (nickdesaulniers) [reply common/gic.h]: thanks, changed from `core + 1` to `BIT(core)`.
